### ISSUE TITLE
VMware compatibility with WSL2 (should raise awareness of limitations)

### ISSUE THREAD

**Issue by polarathene** [MicrosoftDocs/WSL#1820]
### Documentation Issue

> _Note_
> _WSL 2 will work with [VMware 15.5.5+](https://blogs.vmware.com/workstation/2020/05/vmware-workstation-now-supports-hyper-v-mode.html) and [VirtualBox 6+](https://www.virtualbox.org/wiki/Changelog-6.0). Learn more in our [FAQs.](https://learn.microsoft.com/en-us/windows/wsl/faq#will-i-be-able-to-run-wsl-2-and-other-3rd-party-virtualization-tools-such-as-vmware--or-virtualbox-)_
> [Source](https://learn.microsoft.com/en-us/windows/wsl/compare-versions#comparing-wsl-1-and-wsl-2)

- [VMware link provided](https://blogs.vmware.com/workstation/2020/01/vmware-workstation-tech-preview-20h1.html) has no mention of WSL, just Hyper-V?
- According to [WSL Wikipedia](https://en.wikipedia.org/wiki/Windows_Subsystem_for_Linux), WSL2 was released in Windows 10 version 2004 (_months later from the blog post_) and backported. Looks like it referenced a preview build of Windows that would have introduced WSL2 though.

> _For those using VirtualBox, you may need to consider the version that you are running and whether it is compatible with WSL 2. (See [WSL GitHub repository issue 798](https://github.com/MicrosoftDocs/WSL/issues/798) for the full discussion. It appears that VirtualBox v6.1.16 works well with WSL 2, but that other versions may experience issues.)_
> [Source](https://learn.microsoft.com/en-us/windows/wsl/compare-versions#exceptions-for-using-wsl-1-rather-than-wsl-2)

Similar caveat for VMware might be relevant?
- Existing VM guests fail to resume if they've toggled CPU features such as `VT-x` / `AMD-V`.
- Those features need to be turned off, which requires losing any suspended state (_[resume not supported](https://docs.vmware.com/en/VMware-Workstation-Player-for-Windows/17.0/com.vmware.player.win.using.doc/GUID-FF4870F3-8166-4EB4-842B-6F94DBA70758.html) when VMware is changed from traditional to Host VBS mode_); or another workaround (_disabling/uninstalling WSL2 / HyperV which requires a system restart to be able to use VMware as before?_).
- [No nested virtualization support](https://docs.vmware.com/en/VMware-Workstation-Player-for-Windows/17.0/com.vmware.player.win.using.doc/GUID-778DD0ED-6FC9-415F-B071-0CEC127BE719.html) (_the [hypervisor itself is running via nested virtualization](https://learn.microsoft.com/en-us/windows/wsl/troubleshooting#error-0x80370102-the-virtual-machine-could-not-be-started-because-a-required-feature-is-not-installed) under Hyper-V?_).

> _Some 3rd party applications cannot work when Hyper-V is in use, which means they will not be able to run when WSL 2 is enabled, such as VMware and VirtualBox._
> _However, recently both VirtualBox and VMware have released versions that support Hyper-V and WSL2. Learn more about [VirtualBox's changes here](https://www.virtualbox.org/wiki/Changelog-6.0) and [VMware's changes here](https://blogs.vmware.com/workstation/2020/01/vmware-workstation-tech-preview-20h1.html)._
> _For troubleshooting issues, take a look at the [VirtualBox issue discussions in the WSL repo on GitHub](https://github.com/MicrosoftDocs/WSL/issues?q=is%3Aissue+virtualbox+sort%3Acomments-desc)._
> [Source](https://learn.microsoft.com/en-us/windows/wsl/faq#will-i-be-able-to-run-wsl-2-and-other-3rd-party-virtualization-tools-such-as-vmware--or-virtualbox-)

- "recently" may be inaccurate if this was referring to 2020.
- "support" is also not seamless (_at least with VMware_) as described earlier.

It'd be helpful to clarify this restricted compatibility upfront, I had checked the docs prior to installing WSL2, which implied I could still access my VMWare guests without configuration changes. If I had known in advance it would have been easier to prepare for, instead of the added friction imposed after enabling Hyper-V / WSL2.

> _Make sure that the hypervisor launch is enabled in your boot configuration. You can validate this by running (elevated powershell):_
> `bcdedit /enum | findstr -i hypervisorlaunchtype`
> _If you see `hypervisorlaunchtype Off`, then the hypervisor is disabled. To enable it run in an elevated powershell:_
> `bcdedit /set hypervisorlaunchtype Auto`
> 
> _Additionally, if you have 3rd party hypervisors installed (Such as VMware or VirtualBox) then please ensure you have these on the latest versions which can support HyperV ([VMware 15.5.5+](https://blogs.vmware.com/workstation/2020/05/vmware-workstation-now-supports-hyper-v-mode.html) and [VirtualBox 6+](https://www.virtualbox.org/wiki/Changelog-6.0)) or are turned off._
> [Source](https://learn.microsoft.com/en-us/windows/wsl/troubleshooting#error-0x80370102-the-virtual-machine-could-not-be-started-because-a-required-feature-is-not-installed)

```console
$ bcdedit /enum | findstr -i hypervisorlaunchtype
hypervisorlaunchtype    Auto
```

VMware Workstation Player 17.x (_installed prior to WSL2_).

---

Additional Context:

A [user recently referenced](https://github.com/MicrosoftDocs/WSL/issues/798#issuecomment-1590660864) a SuperUser answer:

> _When WSL2 is enabled, the "Virtual Machine Platform" feature is turned on in Windows._
> 
> _This is a subset of Hyper-V functionality, which enables a type 1 hypervisor._
> _Once this is enabled, Windows itself is actually running inside the hypervisor._
> 
> _As a result, VMWare is a "nested" (type 2) hypervisor attempting to run inside the VMP type 1 hypervisor._
> _In order for this to work, you'll need all of the components involved to support the Nested Virtualization feature._

---

> - If I set the `hypervisorlaunchtype` to `auto`, WSL can run but VMware cannot run my Win 11 virtual machine and report my platform does not support virtualized amd-v/rvi.
> - If I set the `hypervisorlaunchtype` to `off`, my Win 11 virtual machine can run but WSL can't.
> [Source](https://github.com/microsoft/WSL/issues/10229)

### Link to documentation page

From quoted sources above:
- https://learn.microsoft.com/en-us/windows/wsl/compare-versions#comparing-wsl-1-and-wsl-2
- https://learn.microsoft.com/en-us/windows/wsl/compare-versions#exceptions-for-using-wsl-1-rather-than-wsl-2
- https://learn.microsoft.com/en-us/windows/wsl/faq#will-i-be-able-to-run-wsl-2-and-other-3rd-party-virtualization-tools-such-as-vmware--or-virtualbox-
- https://learn.microsoft.com/en-us/windows/wsl/troubleshooting#error-0x80370102-the-virtual-machine-could-not-be-started-because-a-required-feature-is-not-installed

In fairness, [VMware hasn't updated their own docs since 2020Q2](https://docs.vmware.com/en/VMware-Workstation-Player-for-Windows/17.0/com.vmware.player.win.using.doc/GUID-177F1E77-BFFD-485F-90BB-2E45B6B88678.html), only stating support for VMware Workstation 15.x, as if to imply anything newer is also unsupported 🤷‍♂️ 
- [This sub-article on the topic](https://docs.vmware.com/en/VMware-Workstation-Player-for-Windows/17.0/com.vmware.player.win.using.doc/GUID-6BE34174-6376-4689-A64F-A04EC2425032.html) also suggests that VMware Workstation is compatible with Hyper-V host with VT-x / AMD-V, but that is not the case (_at least on Windows Home_).

### Suggested Improvements

VMware and VirtualBox are referenced in docs together, and direct users to a rather long issue (_382 comments_) from July 2020 at: https://github.com/MicrosoftDocs/WSL/issues/798

There's likely too much noise built up over time in that thread. It may be wiser to have two separate issues in 2023 for WSL2 2.0.0 going forward, one for VirtualBox and another for VMware?

It seems to be the preferred approach to delegate information to a Github issue than refine the docs content on the subject presumably?

---

I'd like to request an improvement on the actual docs that at least clarify the situation with VMware support.

In the "Virtual Machine Settings" window for a guest, the CPU section has two checkboxes in the "Virtualization Engine" subsection. Neither can be enabled when WSL2 / Hyper-V is active:

![image](https://github.com/MicrosoftDocs/WSL/assets/5098581/bd333884-0068-42af-a94a-349d9c5576b3)

This can impact performance, which the [VMware docs "Limitations of Host VBS mode"](https://docs.vmware.com/en/VMware-Workstation-Player-for-Windows/17.0/com.vmware.player.win.using.doc/GUID-778DD0ED-6FC9-415F-B071-0CEC127BE719.html) also notes performance will be degraded. VMware will prompt awareness about mitigations also being applied when attempting to run in a nested context and that it'll affect performance further.

Similar reports on performance becoming abysmal [is noted for VirtualBox in it's dedicated issue](https://github.com/MicrosoftDocs/WSL/issues/798#issuecomment-1428515922).

**Comment by absinthewww** [comment]
I've been having issues with VMs on my laptop for months and months. Specifically, they will freeze, hang, mess up resuming, especially after I have put my laptop on sleep and then opened it again. I thought it was my laptop doing some weird power management stuff, but I realised that it only happens when WSL2 is on.

It has to have something to do with all this compatibility stuff. It's ridiculous tbh, WSL2 + VM was sold as seamless but there are all these issues years on.

**Comment by craigloewen-msft** [comment]
@polarathene  is your request here to update the WSL docs to include information on how to run it with VMware to make it compatible? Our goal on the docs is to keep it to WSL specific information. WSL relies on Hyper-V, and [VMware has support for Hyper-V](https://blogs.vmware.com/workstation/2020/05/vmware-workstation-now-supports-hyper-v-mode.html) our recommendation is that any VMware set up information should be found on the VMware docs.

**Comment by polarathene** [comment]
**TL;DR:** Installing WSL2 prevents access to existing VMware guests (_configured with `VT-x` / `AMD-V`_).
- The user must uninstall WSL2 or accept data loss by discarding suspended VM guest state, otherwise the VM will not run.
- WSL2 docs should not mislead the reader, this caveat should be explicitly documented.


---

> @polarathene is your request here to update the WSL docs to include information on how to run it with VMware to make it compatible?

My request is to **warn the reader** that enabling WSL2 as documented may prevent access to their existing VMs.

I have detailed the issues in the original post above, but effectively following the WSL2 docs:
- An existing VMware guest may be configured with features that are not compatible with ULM / VBS mode, that **prevents them starting**.
- Bigger frustration is **suspended VMs**. You **cannot access these** without:
  - Throwing away the suspended state.
  - Alternatively, reverting the WSL2 installation to resume the VM guest and save anything necessary so that it can be shutdown without **data loss** and reconfigured.

---

> WSL relies on Hyper-V, and [VMware has support for Hyper-V](https://blogs.vmware.com/workstation/2020/05/vmware-workstation-now-supports-hyper-v-mode.html)

I am aware. That link you provided is in my original message (_from a quote of the current WSL2 docs_).

This does not mean a user with VMware is going to have a smooth experience when they enable WSL2 and later discover that they have to **stop what they're doing** and backup WSL2 work to restore later, so that they can uninstall WSL2 to backup the incompatible VM guests.

If they were **warned in advance** from WSL2 docs, instead of being **mislead that VMware is compatible** (_as if no friction would be introduced preventing access to existing VMs_), this would have been a much more positive experience for the user.

---

Regardless, even if the user is aware of the limitations of the Hyper-V compatible mode, due to those limitations **performance of the VM guests suffers**.

For a casual user who doesn't enable the virtualization performance features, they're none the wiser, but this is fairly standard. It's the [**same virtualization features** that WSL2 wants](https://learn.microsoft.com/en-us/windows/wsl/troubleshooting#error-0x80370102-the-virtual-machine-could-not-be-started-because-a-required-feature-is-not-installed).

So yes, VMware guests can be configured to work with WSL2 installed, but it's **not without drawbacks**.

<details>
<summary>WSL2 experience (drawbacks)</summary>

My primary interest in WSL2 was curiosity as I have run Linux as a daily driver. My VMware guest running Linux was working well and I used it with Docker, I was curious to try Docker Desktop on Windows and that required WSL2 too.

WSL2 has some convenience, but depending on workload the resource usage can be quite problematic, especially in low-disk space scenarios where WSL2 poses a risk. I would still endorse a VM guest given all the issues and friction experienced with WSL2.

Additionally, my experience with WSL2 was:
- High memory usage based on disk I/O. Building a project that would require little memory but write 3GB would raise memory usage to a similar amount for quite a while even after the build was done. It would slowly be released unless shutting the WSL2 instance down.
- Memory usage in WSL2 would also carry over into similar disk usage. Which like the memory allocation, would be released with the same pattern.
- There is the new VHD sparse feature, but it's still not quite right with Docker Desktop. Something is going on with that integration would still accumulate disk space and require the GUI to explicitly purge the WSL2 data volume, as the cleanup CLI commands would not cover whatever WSL2 specific storage management was happening there. It is known to be resolved with [`export` + `import` of the data vdisk](https://github.com/microsoft/WSL/issues/4699#issuecomment-1740681411) (earlier [detailed comment](https://github.com/microsoft/WSL/issues/4699#issuecomment-1738771703)).
- If the host disk runs out of space because of these WSL2 operations, WSL2 becomes unresponsive until a system reboot. You cannot shut it down. Likewise this locks memory usage (_which will slowly dissipate still, but you cannot kill the related process_), quite problematic since WSL2 is the source of the disk bloat (_especially with the memory + disk I/O buffer using up actual disk_).

<details>
<summary>Off-topic - Windows corruption</summary>

I have features of Windows that no longer work. For example:

```console
$ powercfg /batteryreport /output battery_report.html
The Power Efficiency Diagnostic library (energy.dll) could not be loaded.
```

I'm inclined to believe the out of disk space events were related to that, causing some corruption (_related commands for windows diagnostics to scan/repair corrupted data it manages fails_). However, that could be due to something else as Windows on this laptop has encountered various problems over the past 6 months related to graphics drivers, forced updates (Windows or vendor firmware), power management, etc. 

</details>

</details>

**Comment by Sheridan-Tech** [comment]
Has this been resolved? Is it possible to run both? VMware is more important to our workflow than WSL, which would be nice to have

**Comment by polarathene** [comment]
You can run a VMware guest but it's performance will be degraded like I mentioned above.

I assume WSL2 docs still haven't been updated with the caveats I documented here to warn VMware users in advance? :\

---

I haven't tried updating both WSL2 and VMware to confirm if it's changed since my report though, but I doubt it (_please let me know if the situation changes_). WSL2 is quite nice to have and convenient, but I've not been using VMware since due to my report here, if it was resolved I'd gladly be using both.

I'll likely switch back to Linux as the host when I can, and run Windows in a guest instead.

**Comment by Sheridan-Tech** [comment]
> You can run a VMware guest but it's performance will be degraded like I mentioned above. If I were to pick one or the other I'd stay with VMware, unless you need CUDA compute with linux containers which is the only area it lacks I think?
> 
> I assume WSL2 docs still haven't been updated with the caveats I documented here to warn VMware users in advance? :\

Yeah, these are all the same problems I have been complaining about since they originally released WSL. Its a no go for us for all of the reasons you listed above. We even ran one machine bare metal via KVM to gain access to the GPU for CUDA.

Its really a shame there isn't a workaround to simply enable/disable VT-x while in the host OS. I really prefer having access to my main Windows desktop and its resources while I am working, and being able to take snapshots and clone images has been a lifesaver. 

Unfortunately, I also need it to run on the same machine since I sometimes travel for work and thus need it to be portable and ready to go at a moments notice. I suppose theoretically I could use a high performance SBC or something, but that would just add an additional device and equipment to carry around

**Comment by de-served** [comment]
I have all this issues too. I need WSL2, Docker Desktop and VMWare at good performance. And it was a pain to export-unregister-disable WSL2/HyperV/hypervisor to restore VMWare functionality and stored states of VM's, delete Docker Desktop etc, jumping vice-versa and return to VMWare without WSL2/Docker Desktop support :|

**Comment by heraldodinero** [comment]
WTF ... this is still broken ?!?!??!?!?!

**Comment by linghengqian** [comment]
I don't think the current issue can be resolved because once Hyper-V is enabled, it takes over all virtual machines. This means that running virtual machines in VMware is always slower than using Hyper-V Manager. See https://techdocs.broadcom.com/us/en/vmware-cis/desktop-hypervisors/workstation-pro/17-0/using-vmware-workstation-player-for-windows-17-0/running-workstation-on-a-hyper-v-enabled-host-win/limitations-of-host-vbs-mode-win.html .

> Here is a list of functional limitations of a VMware Workstation Player for Windows VM running in Host VBS Mode:
> - Nested VMs are not supported: x86 virtualization features (Intel VT / AMD-V) are unavailable to a guest running on a Host VBS Mode VM. Therefore such VMs cannot themselves run Windows with Hyper-V or VBS enabled.
> - PMCs are not supported: x86 performance monitoring counters (PMCs) are unavailable.
> - RTM and HLE are not supported: Restricted transactional memory and hardware lock elision capabilities are not available.
> - PKU is not supported: User-mode protection keys capability is not available.
